### PR TITLE
feat: introduce CosimM to store platform features in a cache

### LINKED ISSUE #97: Cache Target Triple Information for Random Instruction Generation
we currently repeatedly query information about the target to generate legal instructions during cosimulation (https://github.com/leanprover/LNSym/pull/95). 

We ought to refactor our random instruction generation to compute this information once, and then reuse cached information.

### PR DESCRIPTION
**Pull request by bollu**
Closes #97 . 

### Description:

We implement a `CosimM` monad that tests platform features via `platform_features.sh` once. These are then used during instruction generation time. e.g. the presence/absence of SHA-512 instructions.

At least locally, I observe a modest decrease in wall clock time:

##### before 

```
real 131.23
user 79.43
sys 67.27
```

##### after 

```
real 126.06
user 75.35
sys 60.94
```

### License:

By submitting this pull request, I confirm that my contribution is
made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Arm/Insts/DPSFP/Crypto_four_reg.lean [hunk 1]
```diff
@@ -34,12 +35,8 @@ def exec_crypto_four_reg (inst : Crypto_four_reg_cls) (s : ArmState) : ArmState
...
 ----------------------------------------------------------------------
 
-def Crypto_four_reg_cls.eor3.rand : IO (Option (BitVec 32)) := do
-  let feat_check ←
-      IO.Process.output
-      { cmd  := "Arm/Insts/Cosim/platform_check.sh",
-        args := #["-f", "sha3"] }
-  if feat_check.exitCode = 1 then
+def Crypto_four_reg_cls.eor3.rand : Cosim.CosimM (Option (BitVec 32)) := do
+  if (← Cosim.sha3?) then
```
Comment (bollu) [inline Arm/Insts/DPSFP/Crypto_four_reg.lean:39]: non mechanical change, please audit.

File: Arm/Insts/DPSFP/Crypto_three_reg_sha512.lean [hunk 2]
```diff
@@ -94,12 +95,8 @@ def exec_crypto_three_reg_sha512
...
 ----------------------------------------------------------------------
 
-def Crypto_three_reg_sha512_cls.sha512.rand : IO (Option (BitVec 32)) := do
-  let feat_check ←
-      IO.Process.output
-      { cmd  := "Arm/Insts/Cosim/platform_check.sh",
-        args := #["-f", "sha512"] }
-  if feat_check.exitCode == 1 then 
+def Crypto_three_reg_sha512_cls.sha512.rand : Cosim.CosimM (Option (BitVec 32)) := do
+  if (← Cosim.sha512?) then
```
Comment (bollu) [inline Arm/Insts/DPSFP/Crypto_three_reg_sha512.lean:99]: non mechanical change, please audit.

File: Arm/Insts/DPSFP/Crypto_two_reg_sha512.lean [hunk 3]
```diff
@@ -50,12 +51,8 @@ def exec_crypto_two_reg_sha512
...
 ----------------------------------------------------------------------
 
-def Crypto_two_reg_sha512_cls.sha512su0.rand : IO (Option (BitVec 32)) := do
-  let feat_check ←
-      IO.Process.output
-      { cmd  := "Arm/Insts/Cosim/platform_check.sh",
-        args := #["-f", "sha512"] }
-  if feat_check.exitCode = 1 then
+def Crypto_two_reg_sha512_cls.sha512su0.rand : Cosim.CosimM (Option (BitVec 32)) := do
+  if (← Cosim.sha512?) then
```
Comment (bollu) [inline Arm/Insts/DPSFP/Crypto_two_reg_sha512.lean:55]: non mechanical change, please audit.

File: Arm/Cosim.lean [hunk 4]
```diff
@@ -267,14 +267,10 @@ def run_n_tests (verbose : Bool) (logPrefix : String)
...
-def run_all_tests (verbose : Bool) (n : Nat) : IO UInt32 := do
+def run_all_tests_go (verbose : Bool) (n : Nat) : Cosim.CosimM UInt32 := do
   IO.println s!"Performing conformance testing..."
-  let machine_check ←
-    IO.Process.output
-    { cmd  := "Arm/Insts/Cosim/platform_check.sh",
-      args := #["-m"] }
   let mut tasks := #[]
-  if machine_check.exitCode == 1 then
+  if ← Cosim.aarch64? then
```
Comment (bollu) [inline Arm/Cosim.lean:273]: Non mechanical change, please review

File: Arm/Insts/Common.lean [hunk 5]
```diff
@@ -27,12 +28,8 @@ the latest version is available at
...
 -/
 partial def GPRIndex.rand (lo := 0) (hi := 31) :
-  IO (BitVec 5) := do
-  let darwin_check ←
-  IO.Process.output
-      { cmd  := "Arm/Insts/Cosim/platform_check.sh",
-        args := #["-d"] }
-  if darwin_check.exitCode == 1 then
+  Cosim.CosimM (BitVec 5) := do
+  if ← Cosim.darwin? then
```
Comment (bollu) [inline Arm/Insts/Common.lean:32]: Non mechanical change, please review

File: Arm/Insts/DPSFP/Crypto_four_reg.lean (hunk 1 above)
Comment (pennyannn) [reply Arm/Insts/DPSFP/Crypto_four_reg.lean:39]: LGTM

File: Arm/Insts/DPSFP/Crypto_three_reg_sha512.lean (hunk 2 above)
Comment (pennyannn) [reply Arm/Insts/DPSFP/Crypto_three_reg_sha512.lean:99]: LGTM

File: Arm/Insts/DPSFP/Crypto_two_reg_sha512.lean (hunk 3 above)
Comment (pennyannn) [reply Arm/Insts/DPSFP/Crypto_two_reg_sha512.lean:55]: LGTM

File: Arm/Cosim.lean (hunk 4 above)
Comment (pennyannn) [reply Arm/Cosim.lean:273]: Sometimes you put parentheses around `← Cosim.aarch64?`. It probably doesn't matter, just an observation for the inconsistency.

File: Arm/Insts/Common.lean (hunk 5 above)
Comment (pennyannn) [reply Arm/Insts/Common.lean:32]: LGTM

File: Arm/Cosim.lean (hunk 4 above)
Comment (bollu) [reply Arm/Cosim.lean:273]: Fixed, I've opted to bing Lean-like, so no parens. Sorry Shilpi! (Don't worry, I love Lisp too 😄 )

**Comment by bollu** [comment]
@shigoel fixed parens.

**Comment by bollu** [comment]
@shigoel merged to synchronize with latest `upstream/master`. I feel this can be merged, and we can improve `CosimM` iteratively to save more data.

File: Arm/Insts/CosimM.lean [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+This is used for cosimulation testing.
+-/
+
+import LeanSAT
+import Arm.BitVec
+import Arm.State
+
+namespace Cosim
+/-- Information about our current platform. -/
+structure Context where private mk ::
```
Comment (shigoel) [inline Arm/Insts/CosimM.lean:16]: This is gorgeous! Thanks, @bollu.

**Review by shigoel — APPROVED** [review]
Nice! Thanks, @bollu!

**Pull request merged**